kv/kvserver: TestReplicaCircuitBreaker_ResolveIntent_QuorumLoss failed

kv/kvserver.TestReplicaCircuitBreaker_ResolveIntent_QuorumLoss [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20868736?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20868736?buildTab=artifacts#/) on release-26.1 @ [02545949f7686df4d72c95db8769c7a4f11cd88c](https://github.com/cockroachdb/cockroach/commits/02545949f7686df4d72c95db8769c7a4f11cd88c):


```
        	            	  |   | Wraps: (3) replica unavailable: (n1,s1):1 unable to serve request to r81:{/Table/Max-\xfa\x00} [(n1,s1):1, (n2,s2):2, next=3, gen=4, sticky=9223372036.854775807,2147483647]: closed timestamp: 1765870895.967922367,0 (2025-12-16 07:41:35); raft status: {"id":"1","term":6,"vote":"1","commit":22,"lead":"0","leadEpoch":"0","raftState":"StateFollower","applied":22,"progress":{},"leadtransferee":"0"}: have been waiting 3.50s for slow proposal HeartbeatTxn [/Table/Max], [txn: 6f07cdd8]
        	            	  |   | Wraps: (4)
        	            	  |   |   | (opaque error wrapper)
        	            	  |   |   | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
        	            	  |   |   | reportable 0:
        	            	  |   |   |
        	            	  |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.replicaUnavailableError
        	            	  |   |   | 	pkg/kv/kvserver/replica_circuit_breaker.go:268
        	            	  |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).replicaUnavailableErrorRLocked
        	            	  |   |   | 	pkg/kv/kvserver/replica_circuit_breaker.go:291
        	            	  |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).replicaUnavailableError
        	            	  |   |   | 	pkg/kv/kvserver/replica_circuit_breaker.go:277
        	            	  |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaCircuitBreaker).tripSync
        	            	  |   |   | 	pkg/kv/kvserver/replica_circuit_breaker.go:97
        	            	  |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaCircuitBreaker).TripAsync.func1
        	            	  |   |   | 	pkg/kv/kvserver/replica_circuit_breaker.go:91
        	            	  |   |   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1
        	            	  |   |   | 	pkg/util/stop/stopper.go:440
        	            	  |   |   | runtime.goexit
        	            	  |   |   | 	src/runtime/asm_arm64.s:1268
        	            	  |   | Wraps: (5) closed timestamp: 1765870895.967922367,0 (2025-12-16 07:41:35); raft status: {"id":"1","term":6,"vote":"1","commit":22,"lead":"0","leadEpoch":"0","raftState":"StateFollower","applied":22,"progress":{},"leadtransferee":"0"}
        	            	  |   | Wraps: (6)
        	            	  |   |   | (opaque error wrapper)
        	            	  |   |   | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
        	            	  |   |   | reportable 0:
        	            	  |   |   |
        	            	  |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).refreshProposalsLocked
        	            	  |   |   | 	pkg/kv/kvserver/replica_raft.go:1668
        	            	  |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).tick
        	            	  |   |   | 	pkg/kv/kvserver/replica_raft.go:1502
        	            	  |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processTick
        	            	  |   |   | 	pkg/kv/kvserver/store_raft.go:786
        	            	  |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker
        	            	  |   |   | 	pkg/kv/kvserver/scheduler.go:412
        	            	  |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2
        	            	  |   |   | 	pkg/kv/kvserver/scheduler.go:314
        	            	  |   |   | runtime.goexit
        	            	  |   |   | 	src/runtime/asm_arm64.s:1268
        	            	  |   | Wraps: (7) have been waiting 3.50s for slow proposal HeartbeatTxn [/Table/Max], [txn: 6f07cdd8]
        	            	  |   | Error types: (1) *markers.withMark (2) *markers.withMark (3) *kvpb.ReplicaUnavailableError (4) *errbase.opaqueWrapper (5) *errutil.withPrefix (6) *errbase.opaqueWrapper (7) *errutil.leafError
        	            	  | Wraps: (5) forced error mark
        	            	  |   | "breaker open"
        	            	  |   | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	  | Wraps: (6) encountered poisoned latch /Local/Range/Table/Max/Transaction/"6f07cdd8-6151-4530-acc6-9c45b8dbb06c"@0,0
        	            	  | Error types: (1) *kvpb.ReplicaUnavailableError (2) *errbase.opaqueWrapper (3) *errutil.withPrefix (4) *secondary.withSecondaryError (5) *markers.withMark (6) *poison.PoisonedError
        	            	Wraps: (5) error=replica unavailable: (n1,s1):1 unable to serve request to r81:{/Table/Max-\xfa\x00} [(n1,s1):1, (n2,s2):2, next=3, gen=4, sticky=9223372036.854775807,2147483647]: closed timestamp: 1765870895.967922367,0 (2025-12-16 07:41:35); raft status: {"id":"1","term":6,"vote":"1","commit":22,"lead":"0","leadEpoch":"0","raftState":"StateFollower","applied":22,"progress":{},"leadtransferee":"0"}: encountered poisoned latch /Local/Range/Table/Max/Transaction/"6f07cdd8-6151-4530-acc6-9c45b8dbb06c"@0,0 [exhausted] (last error: failed to send RPC: sending to all replicas failed; last error: DRPC: failed to connect to n2 at 127.0.0.1:42371: initial connection heartbeat failed: dial tcp 127.0.0.1:42371: connect: connection refused)
        	            	Error types: (1) *kvpb.AmbiguousResultError (2) *errbase.opaqueWrapper (3) *secondary.withSecondaryError (4) *secondary.withSecondaryError (5) *errutil.leafError
    panic.go:615: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f53f424fd3523210e5809be9ed06c425/logTestReplicaCircuitBreaker_ResolveIntent_QuorumLoss210385011
--- FAIL: TestReplicaCircuitBreaker_ResolveIntent_QuorumLoss (8.65s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaCircuitBreaker_ResolveIntent_QuorumLoss.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-57928

--- a/pkg/kv/kvserver/client_replica_circuit_breaker_test.go
+++ b/pkg/kv/kvserver/client_replica_circuit_breaker_test.go
@@ -8,6 +8,7 @@ package kvserver_test
 import (
 	"context"
 	"fmt"
+	"strings"
 	"sync/atomic"
 	"testing"
 	"time"
@@ -1351,10 +1352,30 @@ func (*circuitBreakerTest) IsBreakerOpen(err error) bool {
 	// NB: we will see AmbiguousResultError here when proposals are inflight while
 	// the breaker trips. These are wrapping errors, so the assertions below will
 	// look through them.
-	if !errors.Is(err, circuit.ErrBreakerOpen) {
-		return false
+	if errors.Is(err, circuit.ErrBreakerOpen) &&
+		errors.HasType(err, (*kvpb.ReplicaUnavailableError)(nil)) {
+		return true
 	}
-	return errors.HasType(err, (*kvpb.ReplicaUnavailableError)(nil))
+	// Unfortunately, selectBestError in dist_sender.go can "flatten" an
+	// AmbiguousResultError by formatting it as a string via NewAmbiguousResultErrorf,
+	// which loses the error markers. In that case, the ReplicaUnavailableError and
+	// ErrBreakerOpen are only present as stringified text, not as proper error types.
+	// See: https://github.com/cockroachdb/cockroach/issues/159582
+	//
+	// TODO(kvserver): Consider fixing selectBestError to properly wrap errors instead
+	// of formatting them as strings, e.g.:
+	//   return kvpb.NewAmbiguousResultError(errors.WithSecondaryError(ambiguousErr, lastAttemptErr))
+	//
+	// For now, we fall back to string matching for this edge case. The "replica
+	// unavailable" text comes from the stringified ReplicaUnavailableError which
+	// is only produced by the circuit breaker code path.
+	if errors.HasType(err, (*kvpb.AmbiguousResultError)(nil)) {
+		errStr := err.Error()
+		if strings.Contains(errStr, "replica unavailable") {
+			return true
+		}
+	}
+	return false
 }
 
 func (cbt *circuitBreakerTest) RequireIsBreakerOpen(t *testing.T, err error) {

